price-feeder/oracle/provider/crypto.go

```diff
@@ -0,0 +1,536 @@
+	if err != nil {
+		p.logger.Warn().Err(err).Msg("crypto: failed to parse candle close")
+	}
+	volume, err := coin.NewDecFromFloat(candlePair.Volume)
+	if err != nil {
+		p.logger.Warn().Err(err).Msg("crypto: failed to parse candle volume")
+	}
+	candle := types.CandlePrice{
+		Price:  close,
+		Volume: volume,
+		// convert seconds -> milli
+		TimeStamp: SecondsToMilli(candlePair.Timestamp),
+	}
+
+	staleTime := PastUnixTime(providerCandlePeriod)
+	candleList := []types.CandlePrice{}
+	candleList = append(candleList, candle)
+
+	for _, c := range p.candles[symbol] {
+		if staleTime < c.TimeStamp {
+			candleList = append(candleList, c)
+		}
+	}
+
+	p.candles[symbol] = candleList
+}
+
+func (p *CryptoProvider) handleWebSocketMsgs(ctx context.Context) {
+	reconnectTicker := time.NewTicker(cryptoReconnectTime)
+	defer reconnectTicker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-time.After(defaultReadNewWSMessage):
+			messageType, bz, err := p.wsClient.ReadMessage()
+			if err != nil {
+				// If some error occurs, check if connection is alive
+				// and continue to try to read the next message.
+				p.logger.Err(err).Msg("failed to read message")
+				if err := p.ping(); err != nil {
+					p.logger.Err(err).Msg("failed to send ping")
+					if err := p.reconnect(); err != nil {
+						p.logger.Err(err).Msg("error reconnecting websocket")
+					}
+				}
+				continue
+			}
+
+			if len(bz) == 0 {
+				continue
+			}
+
+			p.messageReceived(messageType, bz, reconnectTicker)
+
+		case <-reconnectTicker.C:
+			if err := p.reconnect(); err != nil {
+				p.logger.Err(err).Msg("error reconnecting")
+			}
+		}
+	}
+}
+
+// reconnect closes the last WS connection then create a new one and subscribes to
+// all subscribed pairs in the ticker and candle pairs. If no ping is received
+// within 1 minute, the connection will be disconnected. It is recommended to
+// send a ping for 10-20 seconds
+func (p *CryptoProvider) reconnect() error {
+	err := p.wsClient.Close()
+	if err != nil {
+		return types.ErrProviderConnection.Wrapf("error closing Crypto websocket %v", err)
+	}
+
+	p.logger.Debug().Msg("crypto: reconnecting websocket")
+
+	wsConn, resp, err := websocket.DefaultDialer.Dial(p.wsURL.String(), nil)
+	defer resp.Body.Close()
+	if err != nil {
+		return fmt.Errorf("crypto: error reconnect to crypto websocket: %w", err)
+	}
+	p.wsClient = wsConn
+	telemetryWebsocketReconnect(ProviderCrypto)
+
+	return p.subscribeChannels(p.subscribedPairsToSlice()...)
+}
+
+// setSubscribedPairs sets N currency pairs to the map of subscribed pairs.
+func (p *CryptoProvider) setSubscribedPairs(cps ...types.CurrencyPair) {
+	p.mtx.Lock()
+	defer p.mtx.Unlock()
+
+	for _, cp := range cps {
+		p.subscribedPairs[cp.String()] = cp
+	}
+}
+
+// GetAvailablePairs returns all pairs to which the provider can subscribe.
+// ex.: map["ATOMUSDT" => {}, "UMEEUSDC" => {}].
+func (p *CryptoProvider) GetAvailablePairs() (map[string]struct{}, error) {
```

Comment:
this endpoint should return an map with the key as `CurrencyPair.String()` which it is "ATOMUSDT", at the moment it is returning "ATOM_USDT"

So, we need to parse it from `CryptoPairsSummary` to `CurrencyPair` and then `availablePairs[currencyPair.Sting()] = struct{}{}`